RavenDB-18409 : add schema upgrader to version 6

aviv86 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18409

### Additional description

- added schema upgrade from 52 to 60, which will insert index-values to the new 'BucketAndEtag' index that was added to  database related table schemas
- moved all database related table-schemas to dedicated static classes and spited the schemas to 2 versions (`base` and `base60`), with `Current` property pointing to the latest version
-  schema upgraders will use a specific table-schema version,  all other places will use the `Current` property
- these changes fix the NREs that we had in schema-upgrade related tests
- still missing new tests for this, created separate ticket : https://issues.hibernatingrhinos.com/issue/RavenDB-18448

### Type of change

- Bug fix
- New feature

### How risky is the change?

- Moderate 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

aviv86 committed:
RavenDB-18409 : add a schema upgrader to v6.0 which will insert index-values for all the new 'BucketAndEtag' dynamic indexes

aviv86 committed:
RavenDB-18409 : table-schema usage changes :
- move all database related table-schemas to dedicated static classes
- split schemas to 2 versions (base and base60)
- use 'Current' property everywhere expect in schema-upgraders

aviv86 committed:
RavenDB-18409 : change old schema upgrades to use old schema version

aviv86 committed:
RavenDB-18409 : change latest schema upgrader to use the new schema version + cleanup

arekpalinski commented on src/Raven.Server/Storage/Schema/Updates/Documents/60000/From50002.cs:
```diff
@@ -0,0 +1,104 @@
+
+                    var items = fixedSizeIndex == null
+                        ? table.SeekByPrimaryKey(Slices.BeforeAllKeys, skip)
+                        : table.SeekForwardFrom(fixedSizeIndex, 0, skip);
```
We're always doing the seek to the first item and then skipping. The implementation of the skip operation might be actually pretty expensive when we're skipping to the records at the end because we'll iterate over the whole tree. Please check the skip impl:

https://github.com/ravendb/ravendb/blob/3d10ece8670a53f350b10db7460240f27219a345/src/Voron/Data/Fixed/FixedSizeIterators.cs#L317-L372

We should be able to Seek to close to the last record where we ended (optionally skip few item if needed to be on a good position) and then continue the process.

arekpalinski replied:
In general Skip is supposed to deal with relatively small number of items. Not to skip millions of records.

karmeli87 replied:
So we simply need to keep the last processed key?

aviv86 replied:
fixed

aviv86 committed:
RavenDB-18409 : InsertIndexValuesFor adjustments :
- use fixed size Etag indexes to read items starting from last-etag + 1, instead of always starting from the beginning and skipping
- re-read index tree after transactions renewal
- keep track of total processed items since last commit and commit if it reached batch limit

arekpalinski commented on src/Raven.Server/Storage/Schema/Updates/Documents/60000/From50002.cs:
```diff
@@ -0,0 +1,123 @@
+
+                        indexTree = step.WriteTx.ReadTree(indexName, isIndexTree: true);
+
+                        lastEtag += processedInCurrentTx;
```
But `processedInCurrentTx` isn't the etag value. It's just an integer being incremented during the process.

aviv86 replied:
right, thanks
fixed

aviv86 committed:
RavenDB-18409 : address PR comment - fix wrong last-etag calculation

ayende commented on src/Raven.Server/Documents/Schemas/Attachments.cs:
```diff
@@ -0,0 +1,84 @@
+                Slice.From(ctx, AttachmentsTombstones, ByteStringType.Immutable, out AttachmentsTombstonesSlice);
+            }
+
+            DefineIndexesForAttachmentsSchema(AttachmentsSchemaBase);
+            DefineIndexesForAttachmentsSchemaBase60();
```
I like this approach, it will ensure that we have proper way to track changes over time

ayende commented on src/Raven.Server/Storage/Schema/Updates/Documents/60000/From50002.cs:
```diff
@@ -0,0 +1,132 @@
+            InsertIndexValuesFor(step, AttachmentsSchemaBase60, AttachmentsBucketAndEtagSlice,
+                fixedSizeIndex: AttachmentsSchemaBase60.FixedSizeIndexes[AttachmentsEtagSlice],
+                etagPosition: (int)AttachmentsTable.Etag,
+                tableName: AttachmentsMetadataSlice.ToString());
```
Why are you passing this only here?

aviv86 replied:
Attachments are handled differently.. 
I'm not sure why, but they have no indexes that are marked as global
and I wasn't able to read them using `table = new Table(schema, transaction)` overload
see also here : https://github.com/ravendb/ravendb/blob/v5.3/src/Raven.Server/Documents/AttachmentsStorage.cs#L116

ayende commented on src/Raven.Server/Storage/Schema/Updates/Documents/60000/From50002.cs:
```diff
@@ -0,0 +1,132 @@
+
+        private void InsertIndexValuesFor(UpdateStep step, TableSchema schema, Slice indexName, TableSchema.FixedSizeKeyIndexDef fixedSizeIndex, int etagPosition, string tableName = null)
+        {
+            var fst = step.ReadTx.FixedTreeFor(fixedSizeIndex.Name, sizeof(long));
+            if (fst.NumberOfEntries == 0)
+                return;
+
```
Why do you have this? It is that handle repeated updates?
If so, that isn't valid, that is handled by the schema version marker, not the existence of the tree.

aviv86 replied:
this is the index we're using to iterate over the items, e.g. `AllDocsEtags`
if it's empty then there are no items to process

arekpalinski replied:
What about repeated updates that Oren asked about? Will the update work correctly if the schema update will be aborted midway though and we'll start the server again?

ayende replied:
Let's remove this code. Doing the `foreach` on empty tree gives us the same behavior, and easier to follow.

Aborted midway should be fine, we'll overwrite existing entries in this manner.

aviv86 replied:
tested manually, works fine

arekpalinski replied:
Cool
